kroxylicious-kms-provider-aws-kms/src/test/java/io/kroxylicious/kms/provider/aws/kms/AwsKmsServiceTest.java

```diff
@@ -27,9 +35,42 @@ void afterEach() {
         Optional.ofNullable(awsKmsService).ifPresent(AwsKmsService::close);
     }
 
+    @Test
+    @SuppressWarnings("unchecked")
+    void initializeCreatesCredentialsProvider() {
+        // Given
+        var cpc = mock(CredentialsProviderConfig.class);
+        var cp = mock(CredentialsProvider.class);
+        var config = new Config(URI.create("https://host.invalid"), cpc, "us-east-1", null);
+        when(cpc.createCredentialsProvider()).thenReturn(cp);
+
+        // When
+        awsKmsService.initialize(config);
+
+        // Then
+        verify(cpc).createCredentialsProvider();
+    }
+
     @Test
     void detectsMissingInitialization() {
         assertThatThrownBy(() -> awsKmsService.buildKms())
                 .isInstanceOf(NullPointerException.class);
     }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void credentialsProviderLifecycle() {
+        // Given
+        var cpc = mock(CredentialsProviderConfig.class);
+        var cp = mock(CredentialsProvider.class);
+        var config = new Config(URI.create("https://host.invalid"), cpc, "us-east-1", null);
+        when(cpc.createCredentialsProvider()).thenReturn(cp);
```

Comment:
This pair of tests makes me wonder if we have the life cycle quite right as it verifies that credentials provider is invoked but the provider is not involved in the close. So it effectively assumes that the credentials provider can't be cached and each provider is independent. 

Probably reasonable assumptions to be fair.

Reply:
I not really following.  My intent was to add a test to to ensure that once the service is initialised, that its credential provider gets notified when the service is closed.

Reply:
>once the service is initialised, that its credential provider gets notified when the service is closed.

Yes that I agree with, and makes me realise where I was going wrong. The cpc is the config! So its entirely reasonable that its not notified when closing. I was reading it as a factory not a "one shot" rendering of the config to a provider. My bad.